### PR TITLE
Rehash Existing Instances to Prevent Instance History Corruption

### PR DESCRIPTION
**Pull request by PJEstrada**
## Context:
When the hashing algorithm was changed, we were not able to differentiate between a new hash from a user edit, or a new hash from the new algorithm version. 

Now we rehash the entire existing instance_list to always have the latest version of the hash in place.

- Added a test case to replicate this issue
- Pending some more unit tests to check instance history is not getting corrupted.

### CODE REVIEW

File: shared/annotation.py [hunk 1]
```diff
@@ -642,12 +643,30 @@ def detect_and_remove_collisions(self, instance_list):
...
+            new_hash = instance.hash
+            if prev_hash != new_hash:
+                logger.warning('Warning: Hashing algorithm upgrade Instance ID: {} has changed \n from: {} \n to: {}'.format(
+                    instance.id,
+                    prev_hash,
+                    new_hash
+                ))
+                self.system_upgrade_hash_changes.append([prev_hash, new_hash])
+            result.append(instance)
+            self.session.add(instance)
```
Comment (anthony-sarkis) [inline shared/annotation.py:667]: Should we only add to session if the change as occurred?  Context that we use that session add control in other contexts later in the saving process / if there's no changes there's no need to add it yet?

File: shared/annotation.py [hunk 2]
```diff
@@ -642,12 +643,30 @@ def detect_and_remove_collisions(self, instance_list):
...
         return result
 
+    def rehash_existing_instances(self, instance_list):
+        result = []
+        for instance in instance_list:
+            prev_hash = instance.hash
+            instance.hash_instance()
+            new_hash = instance.hash
+            if prev_hash != new_hash:
+                logger.warning('Warning: Hashing algorithm upgrade Instance ID: {} has changed \n from: {} \n to: {}'.format(
```
Comment (anthony-sarkis) [inline shared/annotation.py:660]: Super small suggestion - I'm ok either way
I noticed we get some support tickets where people question the memory `warning` thing. If the default use case here is system upgrades, should we do `info` maybe?

**Comment by anthony-sarkis** [comment]
Great work Pablo! I think your idea of doing the refresh on the new instances was really excellent here. And I like the implementation and the tests.  Good going!

File: shared/annotation.py (hunk 2 above)
Comment (PJEstrada) [reply shared/annotation.py:660]: Sure Anthony! That's a good idea I guess is less invasive for end users that way :)

File: shared/annotation.py (hunk 1 above)
Comment (PJEstrada) [reply shared/annotation.py]: Thanks Anthony! Added it here: https://github.com/diffgram/diffgram/pull/497/commits/ceccfd0d9250587c4e9e6854cc003dfb80594453

**Comment by PJEstrada** [comment]
> Great work Pablo! I think your idea of doing the refresh on the existing instances was really excellent here. And I like the implementation and the tests. Good going!

Thanks Anthony! Already fixed the things you pointed out :) I'll wait for tests to pass and merge to master when ready

**Review by anthony-sarkis — APPROVED** [review]

**Comment by anthony-sarkis** [comment]
Awesome!

**Comment by anthony-sarkis** [comment]
Cool! :)

**Pull request merged**